### ISSUE TITLE
Memoizing AOT Autograd Input Conversion Breaks Training with Tied Parameters

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#85093]
### 🐛 Describe the bug

AOT Autograd converts inputs to a new set of TensorImpls, either by converting them to fake tensor or by detaching the tensors and then adding back in the gradient `x.detach().requires_grad_(x.requires_grad)`.

When that conversion is memoized, so that two tied parameters are converted to the same FakeTensor/Detached Tensor, models with tied parameters fail. You can replicate this failure:

```
diff --git a/functorch/_src/aot_autograd.py b/functorch/_src/aot_autograd.py
index 6bb4401784..74c2577c73 100644
--- a/functorch/_src/aot_autograd.py
+++ b/functorch/_src/aot_autograd.py
@@ -435,8 +435,14 @@ def create_aot_dispatcher_function(
                 # (resnet50_quantized_qat and mobilenet_v2_quantized_qat)
                 # because they use a "running-mean" style op that requires
                 # resizing the running counter buffers stored on the module.
+                seen = {}
+
                 def make_input(x):
-                    return x.detach().requires_grad_(x.requires_grad)
+                    if id(x) in seen:
+                        return seen[id(x)]
+                    out = x.detach().requires_grad_(x.requires_grad)
+                    seen[id(x)] = out
+                    return out
```

and by running in torchdynamo 
```python benchmarks/huggingface.py --ci -d cuda --float32 --backend=aot_nop --training --only=RobertaForCausalLM```

We would like to memoize the the conversion and reflect shared identities of Parameters so that inplace metadata mutations are reflected on tied parameters. We need to fix the tied parameter gradient issue to do so. 

### Versions

master

cc @Chillee @samdow @kshitij12345 @zou3519 @soumith

**Comment by frgossen** [comment]
This was fixed by #85301 ("Correctly handle duplicate arguments to AOTAutograd"), which introduced input deduplication to prevent gradient double-counting for tied parameters. The deduplication logic now lives in `AOTDedupeWrapper` in `torch/_functorch/_aot_autograd/runtime_wrappers.py`. Closing as completed.

(Analysis done by Claude.)
